api/v1alpha1/obcluster_webhook.go

```diff
@@ -186,6 +185,42 @@ func (r *OBCluster) ValidateUpdate(old runtime.Object) (admission.Warnings, erro
 	} else if !oldCluster.SupportStaticIP() && (oldResource.Cpu != newResource.Cpu || oldResource.Memory != newResource.Memory) {
 		return nil, errors.New("forbid to modify cpu or memory quota of non-static-ip cluster")
 	}
+	if newResource.Memory.Cmp(oldResource.Memory) < 0 {
+		if r.Status.Status != clusterstatus.Running {
+			return nil, errors.New("forbid to shrink memory size of non-running cluster")
+		}
+		conn, err := getSysClient(clt, &obclusterlog, r, oceanbaseconst.OperatorUser, oceanbaseconst.SysTenant, r.Spec.UserSecrets.Operator)
+		if err != nil {
+			return nil, err
+		}
+		defer conn.Close()
+		gvservers, err := conn.ListGVServers(context.Background())
+		if err != nil {
+			return nil, err
+		}
+		var maxAssignedCPU int64
+		var maxAssignedMemory int64
+		var memoryLimitPercent float64
+		for _, gvserver := range gvservers {
+			if gvserver.MemAssigned > maxAssignedMemory {
+				if oldResource.Memory.Value() < gvserver.MemoryLimit {
+					memoryLimitPercent = 0.9
+				} else {
+					memoryLimitPercent = float64(gvserver.MemoryLimit) / oldResource.Memory.AsApproximateFloat64()
+				}
+				maxAssignedMemory = gvserver.MemAssigned
+			}
+			if gvserver.CPUAssigned > maxAssignedCPU {
+				maxAssignedCPU = gvserver.CPUAssigned
+			}
+		}
+		if newResource.Memory.AsApproximateFloat64()*memoryLimitPercent < float64(maxAssignedMemory) {
+			return nil, errors.New("Assigned memory is larger than new memory size")
+		}
+		if newResource.Cpu.Value() > 16 && newResource.Cpu.AsApproximateFloat64() < float64(maxAssignedCPU) {
```

Comment:
16 is the default cpu count set to obcluster but not the actual resource, and this validations is under the condition of memory shrink, but cpu can extend, for example, the obcluster is created with cpu = 2, and the cpu count will be 16 in obcluster, then change it to 4, it will not pass your validation.